### DISCUSSION TITLE
컴포넌트 분리 기준에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by byhhh2** [woowacourse-study/2022-woowahan-bansanghwe#108] [React]
> 작성중 일지도?

# 프론트엔드에서의 컴포넌트

### 컴포넌트에 대한 이해

- 컴포넌트란? 전체의 부분, 요소, 부품
- 엔지니어링 측면에서 컴포넌트? 전체 시스템을 구성하는 하나의 부품 혹은 모듈
- 프론트엔드 엔지니어링 측면에서 컴포넌트? UI를 구성하는 UI 요소
- 비행기에는 200만개 정도의 부품이 들어간다.
    - 컴포넌트가 없다면? 하나의 덩어리로 비행기를 구성할 수 있을까?

### 컴포넌트없이 덩어리로 설계하는 것에 대한 단점

- 의존성을 분리할 수 없다.
    - 새로운 비행기를 만들어야할 때 매번 밑바닥부터 만들어야하기 때문에 같은 구조의 비행기를 다시 만들기가 어렵다. (한번 설계했던 과정이 새로운 비행기를 만드는 데에 도움을 주기 힘들다.)
    - 이미 만들어진 비행기를 재사용해서 새로운 비행기를 만들기 쉽지 않다.

### 분리해서 설계하기

- 복잡한 비행기를 설계하기 위해서는 비행기 설계라는 큰 문제를 작은 문제로 나눌 필요가 있다.
- 예를 들어, 비행기를 꼬리, 날개, 몸체 등(전체의 부분)으로 나눈 후 설계하고 조립하는 방식을 사용한다면 효율적인 설계가 가능하다.

## 🧷 그래서 프론트엔드에서의 컴포넌트란?

- 과거의 웹 사이트는 HTML 만으로 구성되었고 사용자가 새로운 페이지로 이동하면 새로운 HTML 페이지(덩어리)를 서버에서 내려주었다.
    - 이 방법은 페이지의 일부만 변경하고 싶어도 전체 페이지를 새로 내려주어야 한다는 점에서 굉장이 비효율적이였다.
    - 하지만 과거의 HTML은 굉장히 단순했기 때문에 컴포넌트라는 개념이 중요하지 않았다.
    비행기로 치자면 이런 간단한 모형 비행기로 볼 수 있다.
       <img width="150px" src="https://user-images.githubusercontent.com/52737532/193409291-38cf2223-3754-444c-8527-c55883f3d32f.png" />
       ~귀요밍..~

 

        
- 이런 방식은 웹이 복잡해지면 큰 문제가 있는데 우선 웹이 복잡해지고, 사용자 interaction이 많아지면 계속해서 새로운 HTML 페이지를 계속 받아와야 하기 때문에 사용자의 경험이 나빠진다. 이유는 계속 HTML에 대한 요청을 보내고 응답을 받아올 때까지 사용자는 기다리면서 흰 화면을 보고 있어야 하므로 깜빡임을 경험하기 때문이다.
- 또 개발자 측면에서 재사용의 어려움이라는 문제가 있는데 HTML 페이지마다 중복되는 코드들이 계속 분산되어 있어서 유지보수에 매우 취약하다.
- 1990년해 후반에 AJAX가 등장하면서 웹에 대변화가 시작되었는데, 필요한 데이터만을 전달받고 화면에 필요한 일부분만 변경할 수 있게 되면서 웹의 성격이 좀 더 동적으로 바뀌게 되었다. 동적으로 바뀜에 따라 사용자와의 interaction이 증가하였고 웹의 복잡성 또한 높아지게 되었다. (즉, 모형 비행기가 아닌 진짜 비행기와 비슷하게 되었다.)
- 비행기 설계와 마찬가지로 복잡한 웹을 작은 문제로 나누기 시작했다. 복잡한 웹을 작은 문제로 나누게 되면 전체 페이지에 대한 문제가 작은 element group에 대한 문제로 바뀌게 된다. js 통해 element group을 관리할 수 있기 때문에 **프론트엔드의 컴포넌트는 element group을 반환하는 js 함수**라고 정의할 수 있다.
- 컴포넌트를 통해 복잡한 웹을 효율적으로 설계해 나갈 수 있게 되었다. 즉, 컴포넌트를 잘 만든다는 것은 유지보수를 용이하게 한다는 것.

### Atomic Design

- 화학적 개념을 이용해 컴포넌트를 분리하는 방법론
- 컴포넌트 분리 기준에 대해 생각해볼 수 있게 해주는 좋은 인사이트를 제공한다.
- 5단계로 구성

```tsx
ATOMS (원자) 

- 물질을 구성하는 최소 단위의 입자 
- 즉, 유저 인터페이스를 구성하는 최소 단위의 블록 
- 각 원자들은 고유한 특성을 가지고 있다. 그렇기 때문에 서로 조합하여 더 넒은 방식으로 사용될 수 있다. 

예시: <input>, <button> 등과 같은 HTML tag
```

![image](https://user-images.githubusercontent.com/52737532/193409378-39fbfdb3-fec1-4d89-8816-472cf7782ca6.png)

```tsx
MOLECULES (분자) 

- 원자들을 하나의 단위로 동작 시키는 UI 컴포넌트 
- 책임을 가지며, 다른 곳에서 재사용될 수 있다. 

예시: 어떤 목적과 기능도 가지지 않던 <input>, <button> 등과 같은 HTML tag들이 결합하여 
     검색을 하기 위한 UI 컴포넌트가 된다. 
		 검색창은 검색에 대한 책임을 갖고, 다른 곳에서 재사용될 수 있다. 
```

![image](https://user-images.githubusercontent.com/52737532/193409394-1c71a071-836a-482d-a5d3-5fa92672fe80.png)

```tsx
ORGANISMS (유기체) 

- 하나 이상의 책임을 갖거나 반복적인 원자, 분자의 조합 
- 특정 영역을 차지하는 기준에 따라 조합할 수 있다. 

예시: <메인 로고>, <검색창>, <내 정보보기 버튼>들이 존재하는 [헤더 영역]을 유기체로 구성할 수 있다. 
```

![image](https://user-images.githubusercontent.com/52737532/193409404-d0826dde-19ff-4fdb-be4a-b5b088333af5.png)

```tsx
TEMPLATES (템플릿) : 여기서부터 화학적 비유 끝  

- 템플릿은 원자, 분자, 유기체 단계의 컴포넌트들을 배치하는 레이아웃 
- 실제 데이터가 반영되지 않은 상태 

예시: 아직 데이터가 반영되지 않았지만 레이아웃이 잡혀있는 화면  
```

![image](https://user-images.githubusercontent.com/52737532/193409425-3285cd10-6c0c-4b04-8e67-6b7125b1e460.png)

```tsx
PAGES (페이지)

- 실제 데이터가 템플릿에 전달되면 페이지가 완성된다.

예시: 데이터가 반영된 화면 
```

![image](https://user-images.githubusercontent.com/52737532/193409437-7165affb-59af-46a1-9c21-218abfd10562.png)

### Atomic Design의 단점

- 원자 요소에는 태그 뿐만 아니라 스타일 요소도 포함될 수 있다.
- 원자들이 결합되어도 반드시 분자가 되는 것은 아닐 수 있다. (새로운 특성이 항상 생기는 것이 아니기 때문)
- 각 계층으로 어떤 요소들을 분리해야할 지 기준이 모호하다.
- 계층이 5개나 되다보니 구성하려면 많은 시간이 소요되고, 하나의 원자 요소가 변화됐을때 각 계층에 끼치는 사이드 이펙트가 크다.
- Template와 Page 단계는 재사용되기 어렵다.

## 🧷 컴포넌트 어떻게 잘 만들 수 있을까?

- 컴포넌트를 어떻게 하면 잘 설계할까?
- 컴포넌트를 잘 만들면 분리도 쉬워질 것이다!

> **해결하고자 하는 문제**
> : 소프트웨어의 요구사항은 계속 변화한다. 
> : 만약 재사용가능하지 않은 컴포넌트거나 변경에 따른 부수효과가 많으면 리소스가 많이 들어간다. 
> : 컴포넌트 하나를 요구사항 수정에 따라 대대적으로 수정하게 되면 그때마다 개발자가 계속해서 일을 해야한다.  
>  
> 1. 재사용 가능한 컴포넌트 만들기
> 2. 변경에 따른 부수효과 최소화 하기
> 

### 컴포넌트를 만들어보면서 문제점 파악하기

![image](https://user-images.githubusercontent.com/52737532/193409456-97ecfcdd-cd54-41ba-9bf0-0cc6d2340b6d.png)

react를 통해 Input을 만든다고 생각해보자. 

1. input이므로 value가 필요하다. useState를 통해 value를 정의해준다. 
2. input이 사용자에게 어떻게 보여질지 UI를 정의해준다. JSX를 통해 정의해준다. 
3. 마크업과 value를 연결하기 위해 event handler도 정의해준다. 

그런데 새로운 요구사항이 등장한다.

![image](https://user-images.githubusercontent.com/52737532/193409465-9dd360ec-2dee-4afe-9a2e-fafdd470703b.png)

이 전에 만든 컴포넌트를 재사용하기에는 UI가 달라져버렸다. 
하지만, input 자체의 로직은 동일하다. 이를 어떻게 재사용할 수 있을까?
당장 생각나는 방법은?

1. 다른 컴포넌트로 만들기 ⇒ 아까 만들었던 컴포넌트가 재사용이 불가능하다는 것을 입증
2. 스타일을 props로 주어서 분기처리 하기 ⇒ side effect 초래 

이러한 일이 발생하는 이유는 뭘까?
답은 **데이터를 다루는 로직과 UI 로직(마크업)이 하나의 컴포넌트에** 뭉쳐있기 때문 
즉, 하나의 컴포넌트에 두개의 관심사가 뭉쳐있어 발생한 문제.
요구사항이 변하면서 계속적으로 변화(같은 데이터 로직을 갖음)를 요구할텐데 이때마다 계속해서 새로운 컴포넌트를 만들어야 할까? props를 통해 스타일 분기를 해야하나? 

### 이 문제를 Headless라는 개념으로 해결해보자.

> 문제: 하나의 컴포넌트에 두 가지 관심사가 존재한다.
> 

- Headless는 무엇일까요?
    - 머리가 없다.
    - 머리가 뭐고 몸통은 뭘까?
    - 머리 : 컨텐츠를 보여주는 방법
    - 몸 : 컨텐츠 자체
- Headless는 엔지니어링에서만 사용되는 단어는 아니다.
    - Headless 브라우저 : GUI 환경이 아닌 CLI로 실행되는 브라우저 (즉, GUI를 Head로 생각하면 된다.)
    - Headless 커머스 : 프론트엔드와 백엔드를 나눠서 제공하는 시스템 기술 (즉, 프론트엔드는 고객이 대면하는 UI로써 Head)
- 프론트에서의 Headless?
    - Head : UI, 컨텐츠를 보여주는 방법
    - Body: 데이터, 컨텐츠
- 즉, Headless component란 Head(UI)를 버리고, Body(data)만 남긴 컴포넌트

### 이전 Input 컴포넌트를 Headless로 만들기

![image](https://user-images.githubusercontent.com/52737532/193409489-c94d5be9-7f37-4b6b-bdec-ada441fe7c38.png)

남길 부분 ✅

버릴 부분 ❌

### 대표적인 3가지 Headless 패턴

1. **Compound Component 패턴**  
- Compound : 혼합물이라는 뜻

- 컴포넌트 작성!

```jsx
import { createContext, PropsWithChildren, useContext } from 'react';

interface InputValue {
  id: string;
  value: string;
  type: string;
  onChange: () => {};
}

// InputContext : Context API를 통해 컴포넌트 내부에서 공유될 상태(데이터)를 정의 
const InputContext = createContext({} as InputValue);

// InputWrapper : 부모 컴포넌트를 생성하여 InputContext를 공유한다. 
const InputWrapper = ({
  id,
  value,
  type,
  onChange,
  children,
}: PropsWithChildren<InputValue>) => {
  const contextValue = { id, value, type, onChange };

  return (
    <InputContext.Provider value={contextValue}>
      {children}
    </InputContext.Provider>
  );
};

// Input : 자식 컴포넌트, Context API를 통해 필요한 상태를 공유받는다.  
const Input = ({ ...props }) => {
  const { id, value, type, onChange } = useContext(InputContext);

  return (
    <input id={id} value={value} type={type} onChange={onChange} {...props} />
  );
};

const Label = ({ children, ...props }: PropsWithChildren) => {
  const { id } = useContext(InputContext);

  return (
    <label htmlFor={id} {...props}>
      {children}
    </label>
  );
};

// [선택] 자식 컴포넌트들을 부모 컴포넌트의 프로퍼티로 등록한다. 
// 명시적으로 InputWrapper와 Input, Label이 자식관계를 가짐을 명시적으로 알린다. 
InputWrapper.Input = Input;
InputWrapper.Lable = Label;

export default InputWrapper;
```

- 컴포넌트 사용처!

```jsx
const WhereToUse = () => {
  const [name, setName] = useState('');

  const handleChangeName = (event: ChangeEvent<HTMLInputElement>) => {
    setName(event.target.value);
  };

  return (
    <InputWrapper
      id="name"
      value={name}
      type="text"
      onChange={handleChangeName}
    >
      <InputWrapper.Input />
      <InputWrapper.Label>{name}</InputWrapper.Label>
    </InputWrapper>
  );
};

// 사용처에서는 InputWrapper의 하위 컴포넌트들을 자유롭게 볼 수 있고, 
// 위치를 수정하여 마크업을 수정할 수 있다. 
// 즉, 자식 부분을 customizing 할 수 있다. 
```

```jsx
// 만약 Headless 하지 않았다면?

const WhereToUse = () => {
	return (
		<Input /> 
	)
}// 덩어리로 Input 컴포넌트로 사용하게 된다. 
```

1. **Function as Children Component 패턴** 
- children으로 function이 들어온다.
- 자식에 어떤 것이 들어올지 모른다고 가정한다.
- InputHeadless 컴포넌트는 데이터 로직만을 갖는다.
- InputHeadless의 데이터 로직을 자식 함수에 주입한다.

- 컴포넌트 작성!

```jsx
import { useState } from 'react';
import { ChangeEvent } from 'react';

interface FunctionAsChildrenProps {
  children: Function;
}

const FunctionAsChildren = ({ children }: FunctionAsChildrenProps) => {
  const [value, setValue] = useState('');

  const handleChangeValue = (event: ChangeEvent<HTMLInputElement>) => {
    setValue(event.target.value);
  };

  return children({
    value,
    onChange: handleChangeValue,
  });
};

export default FunctionAsChildren;
```

- 사용처!

```jsx
import { ChangeEvent } from 'react';
import FunctionAsChildren from '.';

export const WhereToUse = () => {
  return (
    <FunctionAsChildren>
      {({
        value,
        onChange,
      }: {
        value: string;
        onChange: (event: ChangeEvent<HTMLInputElement>) => void;
      }) => {
        return (
          <>
            <input id="FunctionAsChildren" value={value} onChange={onChange} />
            <label htmlFor="FunctionAsChildren">label</label>
          </>
        );
      }}
    </FunctionAsChildren>
  );
};
```

1. **custom hook 패턴** 
- 데이터 로직을 어느곳에서나 사용할 수 있다.

```jsx
import useInput from './hooks/useInput';

const CustomHookHeadless = () => {
  const { value, onChange } = useInput();

  return (
    <>
      <input value={value} onChange={onChange} />
      <label>label</label>
    </>
  );
};

export default CustomHookHeadless;
```

```jsx
import { ChangeEvent, useState } from 'react';

const useInput = () => {
  const [value, setValue] = useState('');

  const onChange = (event: ChangeEvent<HTMLInputElement>) => {
    setValue(event.target.value);
  };

  return {
    value,
    onChange,
  };
};

export default useInput;
```

## 참고

- [[테코톡] 브랜의 프론트엔드에서 컴포넌트란](https://www.youtube.com/watch?v=2C1tXVHu9XQ)
- [[테코톡] 호프의 프론트엔드에서 컴포넌트](https://www.youtube.com/watch?v=aAs36UeLnTg)

**Comment by airman5573** [comment]
The Legend Movie 👍

**Comment by JUDONGHYEOK** [comment]
컴포넌트란

**Comment by JUDONGHYEOK** [comment]
컴포넌트를 분리하는 무비 만의 기준

**Comment by JUDONGHYEOK** [comment]
많은일을 하는 기준?

**Comment by JUDONGHYEOK** [comment]
컴포넌트 만들면서 적용한 디자인 패턴

**Comment by JUDONGHYEOK** [comment]
확장을 이용한 방식이란?

**Comment by JUDONGHYEOK** [comment]
단점은?

**Comment by JUDONGHYEOK** [comment]
styled component 대신에 emotion을 사용했던 이유?

**Comment by JUDONGHYEOK** [comment]
제어의 역전이란?

**Comment by JUDONGHYEOK** [comment]
제어의 역전 사용해본 경험?

**Comment by JUDONGHYEOK** [comment]
props driling을 해결해본 경험

**Comment by JUDONGHYEOK** [comment]
context api의 문제점

**Comment by JUDONGHYEOK** [comment]
해결하기 위한 노력?

**Comment by JUDONGHYEOK** [comment]
전역상태 관리란?

**Comment by JUDONGHYEOK** [comment]
전역상태관리에 사용한 라이브러리?

**Comment by JUDONGHYEOK** [comment]
리덕스 장단점?

**Comment by JUDONGHYEOK** [comment]
컨텍스트 api보다 좋은점?

**Comment by JUDONGHYEOK** [comment]
하나 예를 들면?

**Comment by JUDONGHYEOK** [comment]
contextapi는 상태관리 도구입니까?

**Comment by JUDONGHYEOK** [comment]
상태란?

**Comment by JUDONGHYEOK** [comment]
리액트에서 setState가 비동기로 동작하는 이유는?

**Comment by JUDONGHYEOK** [comment]
setState가 이뤄지고 리렌더되는 과정

**Comment by JUDONGHYEOK** [comment]
리액트에서는 setState와 state 를 useState로 받아오는데 state를 직접적으로 변경하지 않는 이유

**Comment by JUDONGHYEOK** [comment]
리렌더링조건?

**Comment by JUDONGHYEOK** [comment]
props와 state차이

**Comment by JUDONGHYEOK** [comment]
prop는 그값을 바꿔서는 안되는데 props의 불변성을 지켜야 하는 이유?
